packages/react-on-rails-pro-node-renderer/src/worker/handleIncrementalRenderStream.ts

```diff
@@ -0,0 +1,141 @@
+  };
+  onRenderRequestReceived: (renderRequest: unknown) => Promise<RenderRequestResult> | RenderRequestResult;
+  onResponseStart: (response: ResponseResult) => Promise<void> | void;
+  onUpdateReceived: (updateData: unknown) => Promise<void> | void;
+  onRequestEnded: () => Promise<void> | void;
+}
+
+/**
+ * Handles incremental rendering requests with streaming JSON data.
+ * The first object triggers rendering, subsequent objects provide incremental updates.
+ */
+export async function handleIncrementalRenderStream(
+  options: IncrementalRenderStreamHandlerOptions,
+): Promise<void> {
+  const { request, onRenderRequestReceived, onResponseStart, onUpdateReceived, onRequestEnded } = options;
+
+  let hasReceivedFirstObject = false;
+  const decoder = new StringDecoder('utf8');
+  let buffer = '';
+  let totalBytesReceived = 0;
+
+  try {
+    for await (const chunk of request.raw) {
+      const chunkBuffer = chunk instanceof Buffer ? chunk : Buffer.from(chunk);
+      totalBytesReceived += chunkBuffer.length;
+
+      // Check total request size limit
+      if (totalBytesReceived > MAX_NDJSON_REQUEST_SIZE) {
+        throw new Error(
+          `NDJSON request exceeds maximum size of ${MAX_NDJSON_REQUEST_SIZE} bytes (${Math.round(MAX_NDJSON_REQUEST_SIZE / 1024 / 1024)}MB). ` +
+            `Received ${totalBytesReceived} bytes.`,
+        );
+      }
+
+      const str = decoder.write(chunkBuffer);
+      buffer += str;
+
+      // Check single line size limit (protects against missing newlines)
+      if (buffer.length > MAX_NDJSON_LINE_SIZE) {
+        throw new Error(
+          `NDJSON line exceeds maximum size of ${MAX_NDJSON_LINE_SIZE} bytes (${Math.round(MAX_NDJSON_LINE_SIZE / 1024 / 1024)}MB). ` +
+            `Current buffer: ${buffer.length} bytes. Ensure each JSON object is followed by a newline.`,
+        );
+      }
+
+      // Process all complete JSON objects in the buffer
+      let boundary = buffer.indexOf('\n');
+      while (boundary !== -1) {
+        const rawObject = buffer.slice(0, boundary).trim();
+        buffer = buffer.slice(boundary + 1);
+        boundary = buffer.indexOf('\n');
+
+        if (rawObject) {
+          let parsed: unknown;
+          try {
+            parsed = JSON.parse(rawObject);
+          } catch (err) {
+            const errorMessage = `Invalid JSON chunk: ${err instanceof Error ? err.message : String(err)}`;
+
+            if (!hasReceivedFirstObject) {
+              // Error in first chunk - throw error to stop processing
+              throw new Error(errorMessage);
+            } else {
+              // Error in subsequent chunks - log and report but continue processing
+              const reportedMessage = `JSON parsing error in update chunk: ${err instanceof Error ? err.message : String(err)}`;
+              log.error({ msg: reportedMessage });
+              errorReporter.message(reportedMessage);
```

Comment:
```suggestion
              errorReporter.message(reportedMessage);
```
`errorReporter.message` already includes `log.error`. Same for other places.